### ISSUE TITLE
Preloading/caching issue in Safari

### ISSUE THREAD

**Issue by hauselin** [jspsych/jsPsych#1548]
It looks like Safari reloads images again on each trial even though it has already preloaded stimuli at the beginning of the experiment. I've only tested it with the `image-button-response` plugin though. See discussion #1547.

**Comment by danmyles** [comment]
Also with `html-keyboard-response`. Issue doesn't occur for me when running the script locally (opening index.html in Safari). But does occur when running from cognition.run.

**Comment by jodeleeuw** [comment]
Is this with 6.3 and using the new preload plugin?

**Comment by hauselin** [comment]
It's with both 6.2 and 6.3. I've tested them on cognition.run: [6.2 version](https://m8seoragzv.cognition.run) and [6.3 version](https://ty7doeyux8.cognition.run).

**Comment by danmyles** [comment]
And also with 6.1 (I've tried running across all three). As I understand it cognition.run to implements pre-loading (so I have commented this out of my own script)?

**Comment by danmyles** [comment]
The issue persists with the new preload plugin—possible it's something to do with cognition.run rather the jsPsych? I've submitted a bug report there too :)

**Comment by sankalp35** [comment]
Issue persists for me with 6.3.0, using the preload plugin to load images. Since Prolific does not screen for browsers, it's difficult for me to run the study. Would appreciate some priority on this issue. 

preloading works well with Firefox and Chrome

**Comment by hauselin** [comment]
> Issue persists for me with 6.3.0, using the preload plugin to load images. Since Prolific does not screen for browsers, it's difficult for me to run the study. Would appreciate some priority on this issue.
> 
> preloading works well with Firefox and Chrome

You could try asking participants to use Firefox or Chrome. Most would comply (at least from my study two weeks ago). I also used the following code to check if it's Safari. If so, I made it such that the timeline object contains only one instruction page that asks them to switch to Firefox/Chrome. More info [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Browser_detection_using_the_user_agent).

```
var is_Safari = navigator.userAgent.includes('Safari') && !navigator.userAgent.includes('Chrome');
```

**Comment by jodeleeuw** [comment]
This looks like it might be a clue to solving the issue: https://stackoverflow.com/questions/61566984/async-image-preloading-by-promise-all-not-working-on-non-chrome-browsers

@becky-gilbert we could try adding this decide to preloading functions. I'll work on it when I have some dev time available, but feel free to try it if you beat me to it.

**Comment by maxblomberg** [comment]
Has anyone made any progress in this regard? I am conducting a small cognitive experiment, which is to be completed with a smartphone. To generally exclude everyone with Safari/iOS would be a great pity. I've tried different variations to integrate the img.decode() promise into the preload-plugin, but have failed miserably.

**Comment by jodeleeuw** [comment]
thanks for bumping this @maxblomberg. I've added it to our next milestone, but it'll be at least a month before we get to that.

I think the relevant bit of code to change is in `jspsych.js`:

https://github.com/jspsych/jsPsych/blob/fb2ad412242e60cad5ea9531546158aa70ab9e7b/jspsych.js#L2699-L2718

If you add the .decode() function here it should help.

**Comment by maxblomberg** [comment]
Thanks for tackling this, @jodeleeuw ! However, i don't think it is a Safari problem anymore, but iOS in general. I changed the code part  in `jspsych.js` you pointed out accordingly as described in [stackoverflow](https://stackoverflow.com/questions/61566984/async-image-preloading-by-promise-all-not-working-on-non-chrome-browsers) to

     function preload_image(source){
      return new Promise(res => {
        var img = new Image();
        img.src = source;
        img.onload = (evt) => {
          console.log('loaded data of single image');
          img.decode().then(() => res(img));
          n_loaded++;
          loadfn(img.src);
          if (n_loaded === images.length) {
            finishfn();
          }
        };
      });

      img.onerror = function(e) {
        errorfn({source: img.src, error: e});
      }

      img_cache[source] = img;
      preload_requests.push(img);
    }

This works fine on Android phones (e.g. Chrome/Firefox), however even using Chrome on an iphone leads to the same results that images seem to get reloaded on every trial. However, this might also be due to my dilettante programming skills and hopefully you guys will come to a solution in the future.

**Comment by jirilukavsky** [comment]
Just to help debugging: all browsers on iOS use the same WebKit engine, therefore they should show the same problem.

> WebKit ... primarily used in its Safari web browser, as well as all iOS web browsers

- https://en.wikipedia.org/wiki/WebKit

**Comment by maxblomberg** [comment]
I think I have found the error. It seems to be depending on the location of the retrieved images. 

When using images uploaded directly to [cognition.run](https://cognition.run), Safari and iOS Browsers produce the strange behavior in which they reload the images every trial and thus, delay and flickering occurs. Example here: https://la0n6bl2dx.cognition.run/ (it a small experiment where you are supposed to tap on the friendly looking face).

However, running the same experiment, no flickering (e.g. correct caching) happens when the images are being retrieved from a different website/server, [formr](https://formr.org/) in this case: https://hjkghcyhos.cognition.run/. (The first trial might be delayed since these images are not preloaded).

Maybe that is something you can change, @javidalpe? Could it be an [CORS issue](https://www.jspsych.org/overview/running-experiments/#cross-origin-requests-cors-and-safe-mode)?

Please let me know if this behavior does or does not reproduce for you.

**Comment by becky-gilbert** [comment]
Hi @maxblomberg, so sorry to take so long to get to this issue, and thanks for your debugging help. 

I just tried to reproduce the problem using the two cognition.run links you posted above, using iPad Safari. For me, the first link that you posted (https://la0n6bl2dx.cognition.run/, "VS_local") works fine - all of the images appear immediately after the fixation, including on the first trial. With the second link that you posted (https://hjkghcyhos.cognition.run/, "VS_foreign"), the image borders appear after a delay, and the images don't appear at all (maybe the images were removed from the formr server?). Unfortunately I can't do any further debugging with these tasks because iPads don't provide developer tools and I don't have access to a Mac.

It sounds like what I'm seeing is the opposite of what you described above, i.e. for you, the problem was with the first link, and it was fixed in the second link. Is that right?

**Comment by becky-gilbert** [comment]
Hi @hauselin, sorry for the long delay - I'm not sure if this issue is still relevant for you, but I'm trying to reproduce it now. 

On iPad Safari/Chrome, I'm finding that the images load immediately with your media-cache-preload-6.2 example (https://m8seoragzv.cognition.run/). 

With the 6.3 example (https://ty7doeyux8.cognition.run/), the images appear after a short delay on the first few trials, and then appear immediately after that (since they've all been loaded already after the first few trials). However, I looked at the source code for this this demo and noticed that the preload trial is actually at the _end_ of the experiment timeline, which means that the images aren't actually preloaded before the image presentation trials. So I created the same demo, but with the preload trial at the start: https://jsnzirdb2h.cognition.run/ For me, all of the images appear immediately on iPad Safari/Chrome.

So at this point I'm not able to reproduce the problem. I'm wondering if you (or anyone else) could please check these two links on a Mac laptop/desktop and/or an Apple mobile device (iPad or iPhone)? I'm wondering if I can't reproduce this preloading/caching issue because it only happens on Desktop Safari?
- https://m8seoragzv.cognition.run/ (your 6.2 example)
- https://jsnzirdb2h.cognition.run/ (my 6.3 example)

Thanks very much for your help and patience!

**Comment by becky-gilbert** [comment]
Hi @danmyles, I'm trying to reproduce this issue and haven't had any luck so far. Just wondering if you could provide some demo code and/or a cognition.run link that produces this problem, please? Thanks!

**Comment by hauselin** [comment]
Hi @becky-gilbert, thanks for working on this. I don't use Safari myself and have been asking participants to not use Safari, so I'm not entirely sure if there's still a problem. But I ran your 6.3 example in Safari (see gif below) and it seems like there's still network activity going on on each trial (I don't see such activity in Chrome/Firefox). 

![safari](https://user-images.githubusercontent.com/7620977/125153387-88749480-e110-11eb-9003-b8cb5e8e639c.gif)

**Comment by danmyles** [comment]
Hi @becky-gilbert 

The issue I experienced at the time was that same as @hauselin example above. The image had to re-load from scratch each time it was displayed. I was animating the movement of a slot machine reel, this required introducing several new images over and over as the reel spun. So the issue was a total disaster, in the end I just blocked Safari from loading the experiment and re-directed to a message asking the user to try a different browser.

I just tested our task again now on Safari and the images appears to be displaying without issue. I'm on top of current software updates so maybe Apple has released an update in the intervening weeks that has dealt with the issue?

Here is a demo of the task we developed: https://77ned2rt4n.cognition.run . You will have to skip through the instructions to the stimulus display sequence (just mash space bar).

The issue did not occur on safari for me when running locally, only when the exp was hosted online.

Happy to share the actual script is that helps too LMK.

Hope this helps,

Dan

**Comment by danmyles** [comment]
** I should add that the exp was intended for desktop only (we noted that tablets and smart phones were ineligible via Prolfic).

So I doubt it will display properly on iPad/iPhone.

**Comment by becky-gilbert** [comment]
Hi @danmyles and @hauselin - thanks so much to both of you for getting back to me.

I created a [BrowserStack](https://www.browserstack.com/) account so that I could test on Mac/Safari, and I'm now able to reproduce the problem. On BrowserStack, the problem occurs on older versions of Mac/Safari, such as Catalina/13.1 and Mojavi/12.1. However, the problem does not occur on the latest versions: Big Sur/14.1. This probably explains why I couldn't replicate the issue on my new-ish iPad, which runs Safari 14.1.

@hauselin, I'm not sure what versions of Mac/Safari you were using in the gif above? If you were using Big Sur and Safari 14.1, then there must be factors other than just the OS/browser versions that determine whether this issue occurs.

I also tested the same experiment on a different server, which suggests it's not something specific to cognition.run (although it's still possible that it's caused by something that is common to both servers). 

@maxblomberg as above, I was able to reproduce the image reloading/flickering in your example with earlier versions of Mac/Safari, but it works fine for me with Big Sur/14.1. I'm not able to test your example that hosts the images on formr, because those image links no longer work. But it would be useful to know if that does indeed solve the problem! If possible, would you be able to set up that example again? (Sorry, I know it's been a long time since you posted it). 

In any case, now that I can reproduce the problem, I'll see if it can be fixed using any of the methods @jodeleeuw and @maxblomberg suggested above. 

BTW @danmyles your experiment is super cool! Have you considered sharing the code and posting a note about it on the Discussion board?

**Comment by hauselin** [comment]
I'm using Safari Version 14.0.3 on macOS Catalina (Version 10.15.7).

**Comment by becky-gilbert** [comment]
@hauselin ok thanks!

**Comment by danmyles** [comment]
Thanks @becky-gilbert ☺️!

I'll definitely make all the code available once we get the first paper out. I'd be very happy to share it on the discussion board too! Although it was the first time I'd coded up something in jsPsych so it's probably a bit Frankenstein under the hood...

**Comment by maxblomberg** [comment]
Hi @becky-gilbert,

thanks for tackling and sorry for taking so long. Yes, while cleaning up my experiments in the last months I accidentally deleted the image files from formr. However, I reuploaded them so the links above should work again (VS Local - images hosted on cognition: https://la0n6bl2dx.cognition.run/; VS foreign - images hosted on formr: https://hjkghcyhos.cognition.run/)

I think the MacOS/iOS Update in the meantime solved the issue since the preloading/caching problem no longer reproduces for me, independent of file hosting (cognition or formr), at least for my versions (MacOS Big Sur 11.4; iOS 14.6 - Safari 14.1.1).

But I guess I will stick to the workaround to upload the images somewhere else in order to make the experiment work for people with an older OS. 

Cheers,
Max

**Comment by becky-gilbert** [comment]
@maxblomberg thanks very much for re-uploading your images - being able to test your formr-hosted image example is really helpful for narrowing down the source of the problem. I've just tested your links again and get the same results that you report: on older MacOS/Safari, the images don't reload on each trial in the formr-hosted version, but they do in the cognition-hosted version.

> I think the MacOS/iOS Update in the meantime solved the issue since the preloading/caching problem no longer reproduces for me, independent of file hosting (cognition or formr), at least for my versions (MacOS Big Sur 11.4; iOS 14.6 - Safari 14.1.1).

Ok great, this confirms what I've found so far using BrowserStack.  

> But I guess I will stick to the workaround to upload the images somewhere else in order to make the experiment work for people with an older OS.

Yep I think it's reasonable to expect people to be using older MacOS/iOS versions, which is why I'd still like to see what we can do to fix this.

Thanks again!

**Comment by jodeleeuw** [comment]
Looking over this thread again as it's marked for `7.0`. @becky-gilbert it sounds like the problem is related to how images are hosted/loaded on cognition.run and not something that jspsych is doing. Does that seem right to you?

**Comment by becky-gilbert** [comment]
Sorry for not posting an update on this sooner! Yes, this is related to how the images are hosted - specifically the cache settings in the response header. I'll close this issue now since it's not a jsPsych issue. And I'll post more about the server cache settings that solve this problem ASAP.

**Comment by javidalpe** [comment]
Hello everyone. Thank you for sending your debug sessions. It seems like browsers are not caching the same way. Some browsers are not caching stimuli, new requests are thrown and flickering occurs. 

This is not happening on localhost or other services. I think this could be due to a lack of explicit cache directives from cognition responses. Following @becky-gilbert suggestions, I've configured the server to include Cache-Control headers (https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control) in the response to a stimuli request:
-  First, the redirection response that redirects the browser to the CDN now includes a Cache-Control max-age header following HTTP RFC specification https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.3.3:

> The requested resource resides temporarily under a different URI. Since the redirection might be altered on occasion, the client SHOULD continue to use the Request-URI for future requests. This response is only cacheable if indicated by a Cache-Control or Expires header field.

- Second, the response from the CDN that hosts stimilu includes a Cache-Control max-age header. This only applies to new uploaded stimuli, not previous ones.

I hope this fixes the issue. Even if the redirection is not cached, the final static resource should be cached now.  Looking forward to see your tests :)

**Comment by becky-gilbert** [comment]
Hi all, sorry for the delay in providing an update on this. @javidalpe is correct that the problem seems to be caused by a lack of explicit cache directives combined with different behavior across browsers/versions: 
1. The server's responses to image requests were missing some explicit cache settings - specifically "max-age" in "Cache-Control", "Last Modified" and/or "ETag". All of these settings relate to whether or not the file has changed since it has last been accessed by the browser, and therefore whether it should be loaded from the browser's cache vs requested again from the server. I'm not sure if including any one of these alone is sufficient to fix the problem, because I haven't been able to test comprehensively across browsers/versions and settings.
2. Browsers/versions seem to have different default behavior when this cache information is missing, which is why the re-requesting behavior was only reproducible on older Safari versions.

## TL; DR
If you're having this problem, make sure that the server you're using to host the images responds with a header that contains one or more of the following:
- "max-age" value in "Cache-Control" 
- "Last-Modified" 
- "ETag" 

You can check this by opening the Network tab in the developer tools, loading the page, and inspecting the response header for the image request (see [here for docs with Chrome dev tools](https://developer.chrome.com/docs/devtools/network/reference/#headers)).

## formr
Here is an example of a response header from the formr server, which as you can see includes a "max-age" value in "Cache-Control", as well as "Last-Modified" and "ETag" fields:
![formr_response_header](https://user-images.githubusercontent.com/9041788/132588197-92a9c737-5dd4-4aee-866e-c1b19de182d9.png)
In this case, preloading/caching works as it should: the images are only requested once, and then they're retrieved from the browser cache.

## cognition.run
On cognition.run, the image request is redirected to the CDN location. The cache settings were missing from the initial redirect response, but included in the CDN response:
![cognition_response_header](https://user-images.githubusercontent.com/9041788/132588457-0afc3add-fe31-49dd-90b9-8fc561abefed.png)
In this case, it seems that the redirect response is determining the cache behavior, because the images are still re-requested on every presentation, despite the explicit cache settings in the CDN response. 

However, I then tried replacing the image paths in the experiment code with the CDN URLs, in order to avoid the redirect response. When the images are loaded _directly_ from the CDN, they are only requested once and then load from the browser cache during the experiment.
![cognition_response_header_loadFromCDN](https://user-images.githubusercontent.com/9041788/132589205-e1e71482-0cd3-40c5-b6b3-339d08e8c4b3.PNG)

## JATOS
It so happened that, just like cognition.run, I was also missing all of these cache settings on the separate JATOS/Nginx server that I was using to test this problem. I *think* we've fixed this problem on our server by adding a "max-age" value to "Cache-Control" in the Nginx response header. This "max-age" value is fairly short, since we only need the files to load from the cache for the duration of the experiment. It's also specific to media files, because we want to be sure that the browser is always using the latest versions of JavaScript and other files, and these files aren't loaded _during_ the experiment, so there's no real advantage to caching. 

Ideally we would also include the "Last Modified" or "ETag" fields, since this eliminates the need for an arbitrary "max-age" and ensures that the browser is always using the most up-to-date file versions. However, I don't think this is possible via Nginx alone, because I think "Last Modified"/"ETag" values would need to be passed in from the JATOS application. This means that the JATOS developers would need to make this change on their end. I've [posted about this on the JATOS support forum](https://groups.google.com/g/jatos/c/kevNzKeTQJM), in case anyone is interested in reading/commenting.

**Comment by becky-gilbert** [comment]
Hi again @javidalpe! I just tested this example experiment again: https://la0n6bl2dx.cognition.run/ on MacOS Mojave Safari 12.1.2, and I'm still seeing the flickering image presentations. The network tab is showing that the images are re-requested from the server each time they're displayed, rather than loading from the cache. As I mentioned above, I think the problem is with the response header from the initial request (e.g. https://la0n6bl2dx.cognition.run/img/<file name>) rather than the CDN response header, since replacing the image path with the CDN link in the code seems to fix the problem. I'm also still not seeing a Cache-Control max-age setting in either the redirect or CDN response... Any ideas? Maybe you could test this same experiment link and see if you're getting the same thing?
![cognition_response_header_new](https://user-images.githubusercontent.com/9041788/132594510-a359fd47-20dd-4d8f-b8da-e862280e2f15.png)

**Comment by javidalpe** [comment]
Wow @becky-gilbert, your analysis was awesome! Thank you so much.

You're right. I decided to remove the cache-control header from the redirection. The 302 response redirects the browser to the latest version of the stimuli. This is important to avoid the CDN cache while developing the task. Given a new stimuli uploaded to cognition.run, if we  send a cache-control max age header,  the browser will cache the redirection, the browser will request the same old stimuli version to the CDN,  the old stimuli will be served by the CDN and the developer wont see the new stimuli.

By removing the cache-control, the browser will always request the stimuli to the server. Then cognition.run can redirect the browser to the last version of the stimuli in the CDN.

I'll continue investigating...

**Comment by becky-gilbert** [comment]
Hi @javidalpe, ah, that makes sense. I think the typical solution to this problem is to include "Last-Modified" and/or "ETag", so that the browser can compare the values from the cached file against the server's version in order to decide whether the file has been updated since caching and needs to be requested again. But I don't know if/how this works with the initial redirection... I'll look into it too and let you know if I find anything useful.

Thanks very much for your help in trying to figure this out. Very much appreciated!

**Comment by cainesap** [comment]
Hello,

I think I'm seeing a problem similar to the one described above. I preload my images, but believe they are being newly requested for each trial, based on info in the network tab.

I'm presenting 2 images side by side, and it means that sometimes one image is rendered after the other one, rather than simultaneously as desired. So far I've had this reported with Chrome, Mac OS.

Reading the above, I do see cache-control, max-age and etag in the headers, but wonder if there's something I should be changing in the cache settings for my server all the same?

Sorry, I know this is a closed issue now, so if you prefer I start a new discussion or if there's a more recent one, please do let me know. Possibly also, I should update to 7.0? (I'm using 6.3).

many thanks for any pointers or workarounds, Andrew
